dns_sd.c

```diff
@@ -0,0 +1,295 @@
+
+/* Some functions for handling common linked list operations */
+static void dnssd_remove_node(struct dns_sd_discovery_data **ddata, int n)
+{
+
+	struct dns_sd_discovery_data *d, *ndata, *ldata, *tdata;
+	int i;
+
+	d = *ddata;
+
+	if (n == 0) {
+		tdata = d->next;
+		dnssd_free_discovery_data(d);
+		d = tdata;
+	} else {
+		for (i = 0, ndata = d; ndata->next != NULL; ndata = ndata->next) {
+			if (i == n) {
+				/* Could be NULL or node, both are OK */
+				tdata = ndata->next;
+				/* free the node to be removed */
+				dnssd_free_discovery_data(ndata);
+				/* snip it out */
+				ldata->next = tdata;
+				break;
+			}
+			ldata = ndata;
+			i++;
+		}
+	}
+
+	*ddata = d;
+}
+
+/* The only way to support scan context from the network is when
+ * DNS Service Discovery is turned on
+ */
+
+struct iio_scan_backend_context {
+	struct addrinfo *res;
+};
+
+static int dnssd_fill_context_info(struct iio_context_info *info,
+		char *hostname, char *addr_str, int port)
+{
+	struct iio_context *ctx;
+	char uri[MAXHOSTNAMELEN + 3];
+	char description[255], *p;
+	const char *hw_model, *serial;
+	int i;
+
+	ctx = network_create_context(addr_str);
+	if (!ctx) {
+		ERROR("No context at %s\n", addr_str);
+		return -ENOMEM;
+	}
+
+	if (port == IIOD_PORT)
+		sprintf(uri, "ip:%s", hostname);
+	else
+		sprintf(uri, "ip:%s:%d", hostname, port);
+
+	hw_model = iio_context_get_attr_value(ctx, "hw_model");
+	serial = iio_context_get_attr_value(ctx, "hw_serial");
+
+	if (hw_model && serial) {
+		snprintf(description, sizeof(description), "%s (%s), serial=%s",
+				addr_str, hw_model, serial);
+	} else if (hw_model) {
+		snprintf(description, sizeof(description), "%s %s", addr_str, hw_model);
+	} else if (serial) {
+		snprintf(description, sizeof(description), "%s %s", addr_str, serial);
+	} else if (ctx->nb_devices == 0) {
+		snprintf(description, sizeof(description), "%s", ctx->description);
+	} else {
+		snprintf(description, sizeof(description), "%s (", addr_str);
+		p = description + strlen(description);
+		for (i = 0; i < ctx->nb_devices - 1; i++) {
+			if (ctx->devices[i]->name) {
+				snprintf(p, sizeof(description) - strlen(description) -1,
+						"%s,",  ctx->devices[i]->name);
+				p += strlen(p);
+			}
+		}
+		p--;
+		*p = ')';
+	}
+
+	iio_context_destroy(ctx);
+
+	info->uri = iio_strdup(uri);
+	if (!info->uri)
+		return -ENOMEM;
+
+	info->description = iio_strdup(description);
+	if (!info->description) {
+		free(info->uri);
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+struct iio_scan_backend_context * dnssd_context_scan_init(void)
+{
+	struct iio_scan_backend_context *ctx;
+
+	ctx = malloc(sizeof(*ctx));
+	if (!ctx) {
+		errno = ENOMEM;
+		return NULL;
+	}
+
+	return ctx;
+}
+
+void dnssd_context_scan_free(struct iio_scan_backend_context *ctx)
+{
+	free(ctx);
+}
+
+/*
+ * remove the ones in the list that you can't connect to
+ * This is sort of silly, but we have seen non-iio devices advertised
+ * and discovered on the network. Oh well....
+ */
```

Comment:
Were these "ghost" devices a side effect of publishing iio service with Avahi using the iio.service file and with code in iiod at the same time?

Reply:
No - this happens when the service file is installed with no backing iiod.

Reply:
OK.